src/compiler/checker.ts

```diff
@@ -9544,7 +9604,7 @@ namespace ts {
                     return getUnionType([leftType, rightType]);
                 case SyntaxKind.EqualsToken:
                     checkAssignmentOperator(rightType);
-                    return rightType;
+                    return getRegularTypeOfObjectLiteral(rightType);

```

Comment:
My reasoning was that assigning it to `last` only renders the object "not fresh" if you refer to it via `last`. In this case, you are still returning the object literal, and it has excess properties with respect to `A`.

Maybe we could rationalize it by saying that if we tried to assign it to `last` and it had excess properties relative to `last`, then we would have already given an error. But then again, this seems weird when you consider overload resolution.

Using the same A and B that you've defined:
```ts
declare function foo(param: A): A;
declare function foo(param: B): B;
var last: B;

foo({a: 1, b: 2 }); // returns B
foo(last = {a: 1, b: 2 }); // returns A
```
I don't think taking the regular type here makes sense.